### PR TITLE
[Fastlane.swift] generate Swift API for third party plugins and local actions

### LINKED ISSUE #14940: Plugin how to configure Fastlane.swift
屏幕快照 2019-06-22 下午4 31 04
How to transfer Swift?

### PR DESCRIPTION
**Pull request by joshdholtz**
Related https://github.com/microsoft/fastlane-plugin-appcenter/issues/43
Fixes #14940

### Motivation and Context
Fastlane.swift users have been without actions and plugins for too long 😱

And with the HockeyApp end of life the functionality being transferred to a Microsoft App Center plugin, we need to allow Fastlane.swift users to migrate over if they were using the `hockey` action 🏒

### Description
- `fastlane generate_swift` - only available if a Fastlane.swift project
  - Automatically gets run on `add_plugin` and `update_plugins`
  - User needs to manually run when creating/updating external (local) actions
- Split `SwiftAPIGenerator` into separate generators
  - `SwiftAPIGenerator` is the base class
  - `SwiftFastlaneAPIGenerator` generates the `Fastlane.swift` file of all _fastlane_ actions and global parsing actions
  - `SwiftActionsAPIGenerator` generates the `Actions.swift` file for loaded external (local) actions
  - `SwiftPluginsAPIGenerator` generates the `Plugins.swift` file for third part plugins

#### Demo - Plugins
fastlane-swift-plugins

#### Demo - Actions
fastlane-swift-actions

#### Plugin
1. `bundle exec fastlane init swift`
2. `bundle exec fastlane add_plugin ascii_art`
3. `fastlane/swift/Plugins.swift` should have action for `asciiArt()`
4. Update `fastlane/Fastlane.swift` with a call for `asciiArt()`

#### Actions
1. `bundle exec fastlane init swift`
2. `bundle exec fastlane new_action fromSwift`
3. `bundle exec fastlane generate_swift`
  - Enter `from_swift` as the name
4. `fastlane/swift/Actions.swift` should have action for `fromSwift()`
5. Update `fastlane/Fastlane.swift` with a call for `fromSwift`

### CODE REVIEW

File: fastlane/swift/FastlaneSwiftRunner/FastlaneSwiftRunner.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -205,6 +211,7 @@
 			developmentRegion = English;
 			hasScannedForEncodings = 0;
 			knownRegions = (
+				English,
```
Comment (taquitos) [inline fastlane/swift/FastlaneSwiftRunner/FastlaneSwiftRunner.xcodeproj/project.pbxproj]: 🏴󠁧󠁢󠁥󠁮󠁧󠁿

**Review by taquitos — APPROVED** [review]
THE TIME IS NOW, wow, nice job @joshdholtz! This is what I dreamt of long ago, I'm so happy you came up with a solution! It's been a while since I worked on this part of fastlane, but the changes look totally reasonable to me. I'd love to have somebody else's eyes on it before you merge though 😉
🎈🐐

File: fastlane/lib/fastlane/swift_fastlane_api_generator.rb [hunk 2]
```diff
@@ -13,28 +13,18 @@ def initialize(command_line_tool_name: nil, swift_class: nil, swift_protocol: ni
...
   end
 
-  class SwiftFastlaneAPIGenerator
-    DEFAULT_API_VERSION_STRING = "0.9.1"
-    attr_accessor :tools_option_files
-    attr_accessor :actions_not_supported
-    attr_accessor :action_options_to_ignore
-    attr_accessor :target_output_path
-    attr_accessor :generated_paths # stores all file names of generated files (as they are generated)
+  class SwiftAPIGenerator
```
Comment (mollyIV) [inline fastlane/lib/fastlane/swift_fastlane_api_generator.rb]: Hmm, should `SwiftAPIGenerator` class be here as an empty one? 🤔There's the same one but with the implementation below (line 115).

File: fastlane/swift/Plugins.swift [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+// This autogenerated file will be overwritten or replaced when installing/updating plugins or running "fastlane generate_swift"
+//
+//  ** NOTE **
+//  This file is provided by fastlane and WILL be overwritten in future updates
+//  If you want to add extra functionality to this project, create a new file in a
+//  new group so that it won't be marked for upgrade
+//
+
+import Foundation
```
Comment (mollyIV) [inline fastlane/swift/Plugins.swift]: Thanks a lot for providing Testing Steps 💪 
I was following these:

> Plugin
bundle exec fastlane init swift
bundle exec fastlane add_plugin ascii_art
fastlane/swift/Plugins.swift should have action for asciiArt()
Update fastlane/Fastlane.swift with a call for asciiArt()

however trying to add a different plugin: [`secret_keeper`](https://github.com/mollyIV/fastlane-plugin-secret_keeper). 

image

Both `Actions.swift` and `Plugins.swift` are empty - is there anything that I should first do with the plugin itself? 🤔

I did the manual setup (4) when executing `bundle exec fastlane init swift`.

File: fastlane/swift/Actions.swift [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+// This autogenerated file will be overwritten or replaced when running "fastlane generate_swift"
+//
+//  ** NOTE **
+//  This file is provided by fastlane and WILL be overwritten in future updates
+//  If you want to add extra functionality to this project, create a new file in a
+//  new group so that it won't be marked for upgrade
+//
+
+import Foundation
```
Comment (mollyIV) [inline fastlane/swift/Actions.swift]: I was trying to add a new action:

```
$ bundle exec fastlane new_action swiftSayHello
...
$ bundle exec fastlane generate_swift
```

and this is how the `Actions.swift` looks like:

```
import Foundation
func swift(apiToken: String,
           development: Bool = false) {
  let command = RubyCommand(commandID: "", methodName: "swift", className: nil, args: [RubyCommand.Argument(name: "api_token", value: apiToken),
                                                                                       RubyCommand.Argument(name: "development", value: development)])
  _ = runner.executeCommand(command)
}
func swiftSayHello(apiToken: String,
                   development: Bool = false) {
  let command = RubyCommand(commandID: "", methodName: "swift_say_hello", className: nil, args: [RubyCommand.Argument(name: "api_token", value: apiToken),
                                                                                                 RubyCommand.Argument(name: "development", value: development)])
  _ = runner.executeCommand(command)
}
```

1. Not sure if the default `args` are okay: `api_token` and `development`. 
2. Why is the `swift` method generated? 🤔

**Review by mollyIV — COMMENTED** [review]
@joshdholtz that is :fire: :fire::fire: Thank you so much for implementing these changes! 💪 

I believe that will increase the adaptation of `fastlane swift` in a community 😉

I have been testing your changes and have a few questions though ☺️ 🙇

File: fastlane/lib/fastlane/swift_fastlane_api_generator.rb (hunk 2 above)
Comment (joshdholtz) [reply fastlane/lib/fastlane/swift_fastlane_api_generator.rb]: Ahhhh, I needed to declare it before it was getting used so I stubbed it up at the top and then reopened it up below to make the changes 😛 I should probably make this prettier

File: fastlane/swift/Actions.swift (hunk 4 above)
Comment (joshdholtz) [reply fastlane/swift/Actions.swift]: I wonder if this was due to the way you named your action? 🤔 Actions are usually named like `swift_say_hello` so I'm wondering if there is an issue deeper down somewhere in that logic 🙃 I will look into

Comment (mollyIV) [reply fastlane/swift/Actions.swift]: There was a prompt and I typed 👇 

```
Name of your action: swift_say_hello
```

Comment (mollyIV) [reply fastlane/swift/Actions.swift]: > Not sure if the default args are okay: api_token and development.

They are just default params. The equivalent ruby action is created `fastlane/actions/swift_say_hello.rb` and you should edit `available_options` there. Once you're done with that, just rerun `fastlane generate_swift` to regenerate `Actions.swift` file to match the ones you just updated in the ruby file.

> Why is the swift method generated? 🤔

It looks like I had a mess with the `fastlane/actions/...` catalog and there was `swift_action.rb`, so all good 👍 

Thank you @joshdholtz for explaining it! 🙇

**Review by mollyIV — APPROVED** [review]
👋 @joshdholtz , 

I retested the newest changes and I am happy to tell that the issue when a plugin has more than one action has been fixed, great job 💪 

Approved ✅

**Pull request merged**